bookkeeper-server/src/main/java/org/apache/bookkeeper/client/ForceLedgerOp.java

```diff
@@ -0,0 +1,123 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.bookkeeper.client;
+
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.concurrent.CompletableFuture;
+import org.apache.bookkeeper.common.concurrent.FutureUtils;
+import org.apache.bookkeeper.net.BookieSocketAddress;
+import org.apache.bookkeeper.proto.BookkeeperInternalCallbacks.ForceLedgerCallback;
+import org.apache.bookkeeper.util.SafeRunnable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This represents a request to sync the ledger on every bookie.
+ */
+class ForceLedgerOp extends SafeRunnable implements ForceLedgerCallback {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ForceLedgerOp.class);
+    final CompletableFuture<Void> cb;
+    BitSet receivedResponseSet;
+
+    DistributionSchedule.AckSet ackSet;
+    boolean completed = false;
+    int lastSeenError = BKException.Code.WriteException;
+    ArrayList<BookieSocketAddress> currentEnsemble;
+
+    long currentNonDurableLastAddConfirmed = LedgerHandle.INVALID_ENTRY_ID;
+
+    final LedgerHandle lh;
+
+    ForceLedgerOp(LedgerHandle lh, CompletableFuture<Void> cb) {
+        this.lh = lh;
+        this.cb = cb;
+    }
+
+    void sendForceLedgerRequest(int bookieIndex) {
+        lh.bk.getBookieClient().forceLedger(currentEnsemble.get(bookieIndex), lh.ledgerId, this, bookieIndex);
+    }
+
+    @Override
+    public void safeRun() {
+        initiate();
+    }
+
+    void initiate() {
+
+        // capture currentNonDurableLastAddConfirmed
+        // remember that we are inside OrderedExecutor, this induces a strict ordering
+        // on the sequence of events
+        this.currentNonDurableLastAddConfirmed = lh.pendingAddsSequenceHead;
+        LOG.debug("force {} clientNonDurableLac {}", lh.ledgerId, currentNonDurableLastAddConfirmed);
+
+        // we need to send the request to every bookie in the ensamble
+        this.currentEnsemble = lh.metadata.currentEnsemble;
+        this.ackSet = lh.distributionSchedule.getAckSetForForceLedger();
+        this.receivedResponseSet = new BitSet(
+                lh.getLedgerMetadata().getEnsembleSize());
+        this.receivedResponseSet.set(0,
+                lh.getLedgerMetadata().getEnsembleSize());
+
+        DistributionSchedule.WriteSet writeSet = lh.distributionSchedule.getWriteSetForForceLedger();
+        try {
+            for (int i = 0; i < writeSet.size(); i++) {
+                sendForceLedgerRequest(writeSet.get(i));
+            }
+        } finally {
+            writeSet.recycle();
+        }
+    }
+
+    @Override
+    public void forceLedgerComplete(int rc, long ledgerId, BookieSocketAddress addr, Object ctx) {
+        int bookieIndex = (Integer) ctx;
+
+        if (completed) {
+            return;
+        }
```

Comment:
Given that we are sending force() to all bookies in the ensemble and we won't be completed until we receive a response from all bookies, will we ever enter this condition? Maybe we need to assert here instead of just a return??

Reply:
good catch